File: tests/machine/x/cs/join_multi.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program {
    static void Main() {
        List<Customer> customers = new List<Customer> { new Customer { id = 1, name = "Alice" }, new Customer { id = 2, name = "Bob" } };
        List<Order> orders = new List<Order> { new Order { id = 100, customerId = 1 }, new Order { id = 101, customerId = 2 } };
        List<Item> items = new List<Item> { new Item { orderId = 100, sku = "a" }, new Item { orderId = 101, sku = "b" } };
        List<Result> result = (
    from o in orders
    join c in customers on o.customerId equals c.id
    join i in items on o.id equals i.orderId
    select new Result { name = c["name"], sku = i["sku"] }
).ToList();
        Console.WriteLine("--- Multi Join ---");
        foreach (var r in result) {
            Console.WriteLine(string.Join(" ", new [] { Convert.ToString(r.name), Convert.ToString("bought item"), Convert.ToString(r.sku) }));
        }
    }
    public class Customer {
        public int id;
        public string name;
    }
    
    
    public class Order {
        public int id;
        public int customerId;
    }
    
    
    public class Item {
        public int orderId;
        public string sku;
    }
    
    
    public class Result {
        public string name;
        public string sku;
    }
    
    
    
}

File: tests/machine/x/rust/join_multi.rs
#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
struct Customer {
    id: i32,
    name: &'static str,
}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
struct Order {
    id: i32,
    customerId: i32,
}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
struct Item {
    orderId: i32,
    sku: &'static str,
}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
struct Result {
    name: &'static str,
    sku: &'static str,
}

fn main() {
    let customers = vec![Customer { id: 1, name: "Alice" }, Customer { id: 2, name: "Bob" }];
    let orders = vec![Order { id: 100, customerId: 1 }, Order { id: 101, customerId: 2 }];
    let items = vec![Item { orderId: 100, sku: "a" }, Item { orderId: 101, sku: "b" }];
    let result = { let mut tmp1 = Vec::new();for o in &orders { for c in &customers { if !(o.customerId == c.id) { continue; } for i in &items { if !(o.id == i.orderId) { continue; } tmp1.push(Result { name: c.name, sku: i.sku }); } } } tmp1 };
    println!("--- Multi Join ---");
    for r in result {
        println!("{} {} {}", r.name, "bought item", r.sku);
    }
}
